pkg/database/truncator.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2022 Codenotary Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package database
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/codenotary/immudb/embedded/store"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var (
+	// ErrNoTxBeforeTime is returned when retention period is not reached.
+	ErrRetentionPeriodNotReached = errors.New("retention period not reached")
+)
+
+// Truncator provides truncation against an underlying storage
+// of appendable data.
+type Truncator interface {
+	// Plan returns the transaction upto which the log can be truncated.
+	// When resulting transaction before specified time does not exists
+	//  * No transaction header is returned.
+	//  * Returns nil TxHeader, and an error.
+	Plan(time.Time) (*store.TxHeader, error)
+
+	// Truncate runs truncation against the relevant appendable logs. Must
+	// be called after result of Plan().
+	Truncate(*store.TxHeader) error
+}
+
+func newVlogTruncator(d *db) Truncator {
+	return &vlogTruncator{
+		db:      d,
+		metrics: newTruncatorMetrics(d.name),
+	}
+}
+
+// vlogTruncator implements Truncator for the value-log appendable
+type vlogTruncator struct {
+	db      *db
+	metrics *truncatorMetrics
+}
+
+// Plan returns the transaction upto which the value log can be truncated.
+// When resulting transaction before specified time does not exists
+//  * No transaction header is returned.
+//  * Returns nil TxHeader, and an error.
+// The retentionPeriod time is truncated to the day.
```

Comment:
wondering why retentionPeriod needs to be truncated

Reply:
This is just to keep uniformity on when truncation is done for the entire day, because if we say delete data for n days, time in hours/seconds shouldn't matter else half a day data could be truncated on that day depending on the timestamps